File: reference/python/verify_test_vectors.py
"""Verify spec/test-vectors against the reference decoder.

Reads `manifest.json` from the given directory, decodes every PNG entry,
and checks the result against the expected plaintext. Prints PASS/FAIL
per vector and exits non-zero if any vector fails.
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

from stiger_decode import decode


def run(vectors_dir: Path) -> int:
    manifest_path = vectors_dir / "manifest.json"
    if not manifest_path.is_file():
        print(f"manifest not found: {manifest_path}", file=sys.stderr)
        return 2

    manifest = json.loads(manifest_path.read_text())
    vectors = manifest.get("vectors", [])
    if not vectors:
        print("manifest has no vectors", file=sys.stderr)
        return 2

    failures = 0
    for vec in vectors:
        name = vec["name"]
        png = vectors_dir / vec["png"]
        password = vec.get("password") or None
        expected_mode = vec["mode"]
        expected_plaintext = vec["plaintext_utf8"].encode("utf-8")

        try:
            result = decode(png, password)
        except Exception as e:
            print(f"FAIL  {name}: decode raised {e}")
            failures += 1
            continue

        actual_mode = "stealth" if result.encrypted else "open"
        if actual_mode != expected_mode:
            print(f"FAIL  {name}: mode {actual_mode} != expected {expected_mode}")
            failures += 1
            continue
        if result.data != expected_plaintext:
            print(f"FAIL  {name}: plaintext mismatch")
            print(f"      got:      {result.data!r}")
            print(f"      expected: {expected_plaintext!r}")
            failures += 1
            continue
        print(f"PASS  {name}  ({actual_mode}, {len(result.data)} B)")

    print()
    if failures:
        print(f"{failures} of {len(vectors)} vectors failed")
        return 1
    print(f"all {len(vectors)} vectors passed")
    return 0


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("vectors_dir", type=Path,
                        help="spec/test-vectors directory")
    args = parser.parse_args(argv)
    return run(args.vectors_dir)


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: reference/typescript/verify_test_vectors.ts
// Decode every vector listed in `<vectors-dir>/manifest.json` and compare the
// decoded plaintext to `plaintext_utf8`. Mirrors the Python verify script
// next door (`../python/verify_test_vectors.py`). Exit `0` means this TS
// reference is in sync with the shipping Swift engine via the test vectors.
//
// Usage:  npx tsx verify_test_vectors.ts <vectors-dir>

import { readFileSync } from 'node:fs';
import { resolve, basename } from 'node:path';
import sharp from 'sharp';

import { decode } from './stiger_decode.ts';

type Vector = {
  name: string;
  mode: 'open' | 'stealth';
  png: string;
  password?: string;
  plaintext_utf8: string;
};

type Manifest = { vectors: Vector[] };

async function main() {
  const dir = resolve(process.argv[2] ?? '../../test-vectors');
  const manifest: Manifest = JSON.parse(
    readFileSync(resolve(dir, 'manifest.json'), 'utf8'),
  );

  let pass = 0;
  let fail = 0;

  for (const v of manifest.vectors) {
    const pngPath = resolve(dir, v.png);
    const { data, info } = await sharp(pngPath)
      .ensureAlpha()
      .raw()
      .toBuffer({ resolveWithObject: true });
    const rgba = new Uint8Array(data.buffer, data.byteOffset, data.byteLength);

    const result = await decode(rgba, v.password);
    const ok = result.ok && result.message === v.plaintext_utf8;
    const tag = `${v.name} (${info.width}×${info.height})`;
    if (ok) {
      pass++;
      console.log(`✓ ${tag}`);
    } else {
      fail++;
      console.log(`✗ ${tag}`);
      console.log(`  expected: ${JSON.stringify(v.plaintext_utf8.slice(0, 80))}`);
      if (result.ok) {
        console.log(`  got:      ${JSON.stringify(result.message.slice(0, 80))}`);
      } else {
        console.log(`  failed:   ${result.reason}${result.detail ? ' — ' + result.detail : ''}`);
      }
    }
  }

  console.log(`\n${pass} passed, ${fail} failed (${basename(dir)})`);
  process.exit(fail === 0 ? 0 : 1);
}

main().catch((err) => {
  console.error(err);
  process.exit(2);
});
